src/diffusion_for_multi_scale_molecular_dynamics/generators/adaptative_corrector.py

```diff
@@ -0,0 +1,257 @@
+                record_axl = AXL(
+                    A=axl.A.detach().cpu(),
+                    X=axl.X.detach().cpu(),
+                    L=axl.L.detach().cpu(),
+                )
+                entry[key] = record_axl
+            self.sample_trajectory_recorder.record(key="predictor_step", entry=entry)
+
+        return composition_im1
+
+    def corrector_step(
+        self,
+        composition_i: AXL,
+        index_i: int,
+        unit_cell: torch.Tensor,  # TODO replace with AXL-L
+        cartesian_forces: torch.Tensor,
+    ) -> AXL:
+        r"""Corrector Step.
+
+        Note this does not affect the atom types unless specified with the atom_type_transition_in_corrector argument.
+        Always affect the reduced coordinates and lattice vectors. The prefactors determining the changes in the X and L
+        variables are determined using the sigma normalized score at that corrector step. The relative coordinates
+        update is given by:
+
+        .. math::
+
+            x_i \leftarrow x_i + \epsilon_i * s(x_i, t_i) + \sqrt(2 \epsilon_i) z
+
+        where :math:`s(x_i, t_i)` is the score, :math:`z` is a random variable drawn from a normal distribution and
+        :math:`\epsilon_i` is given by:
+
+        .. math::
+
+            \epsilon_i = 2 \left(r \frac{||z||_2}{||s(x_i, t_i)||_2}\right)^2
+
+        where :math:`r` is an hyper-parameter (0.17 by default) and :math:`||\cdot||_2` is the L2 norm.
+
+        Args:
+            composition_i : sampled composition (atom types, relative coordinates, lattice vectors), at time step i.
+            index_i : index of the time step.
+            unit_cell: sampled unit cell at time step i.  # TODO replace with AXL-L
+            cartesian_forces: forces conditioning the sampling
+
+        Returns:
+            corrected_composition_i : sampled composition, after corrector step.
+        """
+        assert 0 <= index_i <= self.number_of_discretization_steps - 1, (
+            "The corrector step can only be invoked for index_i between 0 and "
+            "the total number of discretization steps minus 1."
+        )
+
+        if index_i == 0:
+            # TODO: we are extrapolating here; the score network will never have seen this time step...
+            sigma_i = (
+                self.noise_parameters.sigma_min
+            )  # no need to change device, this is a float
+            t_i = 0.0  # same for device - this is a float
+            idx = index_i
+        else:
+            idx = index_i - 1  # python starts indices at zero
+            sigma_i = self.noise.sigma[idx].to(composition_i.X)
+            t_i = self.noise.time[idx].to(composition_i.X)
+
+        model_predictions_i = self._get_model_predictions(
+            composition_i, t_i, sigma_i, unit_cell, cartesian_forces
+        )
+
+        # to compute epsilon_i, we need the norm of the score. We average over the atoms.
+        relative_coordinates_sigma_score_norm = (
```

Comment:
The norm in the paper is calculated over all pixels. 

I think you are doing this: 
<img width="418" alt="Screenshot 2024-12-02 at 11 34 09 AM" src="https://github.com/user-attachments/assets/a827848c-8045-4612-a97d-399a0fa2d3d3">

and I think the right way of doing it is this:
<img width="379" alt="Screenshot 2024-12-02 at 11 34 30 AM" src="https://github.com/user-attachments/assets/2a130679-8171-43bc-96fa-bf5b771c68b6">

The paper talks about averaging over the batch dimension; you are averaging over the atoms within a single sample.

<img width="691" alt="Screenshot 2024-12-02 at 11 34 59 AM" src="https://github.com/user-attachments/assets/06b435cb-8cc9-4659-b2cc-e2c79fa196d1">

I don't think we should do this just yet...

Reply:
I wasn't sure what was the best approach to take the norm, so I went with something reasonable. Let's stick to the paper recommendations for now.